File: src/test/java/personal/alexgarland/helloworldmvc/testutil/LogbackEventChecker.java
package personal.alexgarland.helloworldmvc.testutil;

import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import java.util.ArrayList;
import java.util.List;
import org.mockito.ArgumentCaptor;
import org.slf4j.LoggerFactory;
import ch.qos.logback.classic.Level;
import ch.qos.logback.classic.Logger;
import ch.qos.logback.classic.spi.ILoggingEvent;
import ch.qos.logback.classic.spi.LoggingEvent;
import ch.qos.logback.core.Appender;

public class LogbackEventChecker {

  private static final Logger ROOT_LOGGER =
      (Logger) LoggerFactory.getLogger(Logger.ROOT_LOGGER_NAME);

  private static Appender<ILoggingEvent> mockAppender;

  @SuppressWarnings("unchecked")
  public static void setup() {
    mockAppender = mock(Appender.class);
    ROOT_LOGGER.addAppender(mockAppender);
  }

  public static void teardown() {
    ROOT_LOGGER.detachAppender(mockAppender);
  }

  private final List<LogEventDetails> expectedDetails = new ArrayList<LogEventDetails>();

  private final ArgumentCaptor<LoggingEvent> captorLoggingEvent =
      ArgumentCaptor.forClass(LoggingEvent.class);

  public LogbackEventChecker withExpectedEvent(Level level, String message) {
    expectedDetails.add(new LogEventDetails(level, message));
    return this;
  }

  public void verifyCalls() {

    int numCalls = expectedDetails.size();

    verify(mockAppender, times(numCalls)).doAppend(captorLoggingEvent.capture());

    List<LoggingEvent> actualEvents = captorLoggingEvent.getAllValues();

    for (int i = 0; i < numCalls; i++) {
      ILoggingEvent actual = actualEvents.get(i);
      LogEventDetails expected = expectedDetails.get(i);

      assertThat(actual.getLevel(), is(expected.getLevel()));
      assertThat(actual.getFormattedMessage(), is(expected.getMessage()));
    }

  }

}

File: src/test/kotlin/personal/alexgarland/helloworldmvc/testutil/LogbackEventChecker.kt
package personal.alexgarland.helloworldmvc.testutil

import ch.qos.logback.classic.Level
import ch.qos.logback.classic.Logger
import ch.qos.logback.classic.spi.ILoggingEvent
import ch.qos.logback.classic.spi.LoggingEvent
import ch.qos.logback.core.Appender
import org.hamcrest.CoreMatchers.`is`
import org.junit.Assert.assertThat
import org.mockito.ArgumentCaptor
import org.mockito.Mockito.*
import org.slf4j.LoggerFactory
import java.util.*

class LogbackEventChecker {

    private val expectedDetails = ArrayList<LogEventDetails>()

    private val captorLoggingEvent = ArgumentCaptor.forClass(LoggingEvent::class.java)

    fun withExpectedEvent(level: Level, message: String): LogbackEventChecker {
        expectedDetails.add(LogEventDetails(level, message))
        return this
    }

    fun verifyCalls() {

        val numCalls = expectedDetails.size

        verify<Appender<ILoggingEvent>>(mockAppender, times(numCalls)).doAppend(captorLoggingEvent.capture())

        val actualEvents = captorLoggingEvent.allValues

        for (i in 0 until numCalls) {
            val actual = actualEvents[i]
            val expected = expectedDetails[i]

            assertThat(actual.level, `is`(expected.level))
            assertThat(actual.formattedMessage, `is`(expected.message))
        }

    }

    companion object {

        private val ROOT_LOGGER = LoggerFactory.getLogger(Logger.ROOT_LOGGER_NAME) as Logger

        private var mockAppender: Appender<ILoggingEvent>? = null

        fun setup() {

            @Suppress("UNCHECKED_CAST")
            mockAppender = mock(Appender::class.java) as Appender<ILoggingEvent>

            ROOT_LOGGER.addAppender(mockAppender)
        }

        fun teardown() {
            ROOT_LOGGER.detachAppender(mockAppender)
        }
    }

}
